coreax/metrics.py

```diff
@@ -428,3 +429,200 @@ def sum_weighted_pairwise_distances(
+        \mathbb{R}^p`, the conditional maximum mean discrepancy is given by:
+
+        .. math::
+
+            \text{CMMD}^2(\mathcal{D}^{(1)}, \mathcal{D}^{(2)}) = ||\hat{\mu}^{(1)} -
+            \hat{\mu}^{(2)}||^2_{\mathcal{H}_k \otimes \mathcal{H}_l}
+
+        where :math:`\hat{\mu}^{(1)},\hat{\mu}^{(2)}` are the conditional mean
+        embeddings estimated with :math:`\mathcal{D}^{(1)}` and
+        :math:`\mathcal{D}^{(2)}` respectively, and :math:`\mathcal{H}_k,\mathcal{H}_l`
+        are the RKHSs corresponding to the kernel functions
+        :math:`k: \mathbb{R}^d \times \mathbb{R}^d \rightarrow \mathbb{R}` and
+        :math:`l: \mathbb{R}^p \times \mathbb{R}^p \rightarrow \mathbb{R}` respectively.
+
+        :param dataset_1: The original dataset :math:`\mathcal{D}^{(1)} =
+            \{(x_i, y_i)\}_{i=1}^n` of ``n`` pairs with :math:`x\in\mathbb{R}^d` and
+            :math:`y\in\mathbb{R}^p`, responses should be concatenated after the
+            features
+        :param dataset_2: Dataset :math:`\mathcal{D}^{(2)} = \{(\tilde{x}_i,
+            \tilde{y}_i)\}_{i=1}^n` of ``m`` pairs with :math:`\tilde{x}\in\mathbb{R}^d`
+            and :math:`\tilde{y}\in\mathbb{R}^p`, responses should be
+            concatenated after the features
+        """
+        # Extract and format features and responses from D1 and D2
+        x1 = jnp.atleast_2d(dataset_1[:, : self.num_feature_dimensions])
+        y1 = jnp.atleast_2d(dataset_1[:, self.num_feature_dimensions :])
+        x2 = jnp.atleast_2d(dataset_2[:, : self.num_feature_dimensions])
+        y2 = jnp.atleast_2d(dataset_2[:, self.num_feature_dimensions :])
+
+        # Compute feature kernel gramians
+        feature_gramian_1 = self.feature_kernel.compute(x1, x1)
+        feature_gramian_2 = self.feature_kernel.compute(x2, x2)
+        cross_feature_gramian = self.feature_kernel.compute(x2, x1)
+
+        # Compute response kernel gramians
+        response_gramian_1 = self.response_kernel.compute(y1, y1)
+        response_gramian_2 = self.response_kernel.compute(y2, y2)
+        cross_response_gramian = self.response_kernel.compute(y1, y2)
+
+        # Invert feature kernel gramians
+        identity_1 = jnp.eye(feature_gramian_1.shape[0])
+        inverse_feature_gramian_1 = coreax.util.invert_regularised_array(
+            array=feature_gramian_1,
+            regularisation_parameter=self.regularisation_parameters[0].item(),
+            identity=identity_1,
+        )
+
+        identity_2 = jnp.eye(feature_gramian_2.shape[0])
+        inverse_feature_gramian_2 = coreax.util.invert_regularised_array(
+            array=feature_gramian_2,
+            regularisation_parameter=self.regularisation_parameters[1].item(),
+            identity=identity_2,
+        )
+
+        # Compute each term in the CMMD
+        term_1 = (
+            inverse_feature_gramian_1
+            @ response_gramian_1
+            @ inverse_feature_gramian_1
+            @ feature_gramian_1
```

Comment:
Does the `inverse_feature @ feature` not simplify to the indentity matrix, or is this becuase we aren't computing a real inverse, but a regularized psuedo-inverse?

Reply:
Yeah because it is a regularised inverse this doesn't just become the identity matrix